File: services/rag/app/secret_scanner.py
"""Pre-ingestion secret scanner.

Wraps Yelp's `detect-secrets` library. Detection is done by its plugin
pipeline (AWS, Azure, Stripe, GitHub, JWT, private keys, and generic
keyword / entropy detectors); the library's default filters already remove
common false positives — UUIDs, indirect references like
``this.config.apiKey``, templated placeholders ``${foo}`` / ``<foo>``,
id-looking strings, lock and swagger files.

On top of that we apply a small post-filter for bareword placeholders
the library still flags (``REDACTED``, ``your-api-key-here``, ``null``,
etc.) — these sneak through because the KeywordDetector treats any value
after ``password:`` / ``secret:`` as a candidate.
"""

import re
import tempfile
from pathlib import Path

from detect_secrets import SecretsCollection
from detect_secrets.settings import transient_settings
from loguru import logger

# Enabled plugins. Keep the list explicit so bumping detect-secrets does
# not silently change behavior. Specific-provider detectors (Stripe,
# GitHub, OpenAI, …) stay first so their typed label shows up in the
# rejection reason when both a specific and generic detector match.
_PLUGINS: list[dict[str, object]] = [
    {"name": "AWSKeyDetector"},
    {"name": "ArtifactoryDetector"},
    {"name": "AzureStorageKeyDetector"},
    {"name": "BasicAuthDetector"},
    {"name": "CloudantDetector"},
    {"name": "DiscordBotTokenDetector"},
    {"name": "GitHubTokenDetector"},
    {"name": "GitLabTokenDetector"},
    {"name": "IbmCloudIamDetector"},
    {"name": "IbmCosHmacDetector"},
    {"name": "JwtTokenDetector"},
    {"name": "MailchimpDetector"},
    {"name": "NpmDetector"},
    {"name": "OpenAIDetector"},
    {"name": "PrivateKeyDetector"},
    {"name": "PypiTokenDetector"},
    {"name": "SendGridDetector"},
    {"name": "SlackDetector"},
    {"name": "SoftlayerDetector"},
    {"name": "SquareOAuthDetector"},
    {"name": "StripeDetector"},
    {"name": "TelegramBotTokenDetector"},
    {"name": "TwilioKeyDetector"},
    {"name": "KeywordDetector"},
    # Default entropy limits: Base64 4.5, Hex 3.0. Raising Base64 further
    # hides real secrets; lowering trips on innocuous identifiers.
    {"name": "Base64HighEntropyString", "limit": 4.5},
    {"name": "HexHighEntropyString", "limit": 3.0},
]

# Bareword placeholders the KeywordDetector still flags. detect-secrets
# has `is_templated_secret` for `${x}` / `<x>` / `{x}` but not for these
# plain-word conventions.
_PLACEHOLDER_VALUES: frozenset[str] = frozenset(
    {
        "",
        "null",
        "none",
        "nil",
        "undefined",
        "true",
        "false",
        "redacted",
        "placeholder",
        "example",
        "changeme",
        "xxx",
        "xxxx",
        "xxxxxxxx",
        "tbd",
        "todo",
        "fixme",
        "n/a",
    }
)

_PLACEHOLDER_PATTERNS: tuple[re.Pattern[str], ...] = (
    # `your-api-key-here`, `my_secret`, `example-token`, etc.
    re.compile(r"^(?:your|my|sample|example|test|fake|dummy|demo)[-_].+", re.I),
    # `...-goes-here`, `...-placeholder`, `...-example`.
    re.compile(r".+[-_](?:here|placeholder|example|sample|value|goes-here)$", re.I),
)

# Member-access chains (`config.apiKey`, `process.env.FOO`, `this.token`).
# KeywordDetector flags `secret: config.apiKey` because it treats the whole
# RHS as a candidate, but a dotted identifier expression is a reference,
# not a literal secret. Require at least one dot so a bare identifier
# (which could itself be a secret) still falls through to detection.
_MEMBER_ACCESS_RE = re.compile(r"^[A-Za-z_$][\w$]*(?:\.[A-Za-z_$][\w$]*)+$")


def _is_noise(secret_value: str | None) -> bool:
    """True when the detected value is a known placeholder, not a real secret."""
    if not secret_value:
        return True
    stripped = secret_value.strip().strip("'\"`")
    if not stripped:
        return True
    lowered = stripped.lower()
    if lowered in _PLACEHOLDER_VALUES:
        return True
    if any(p.match(lowered) for p in _PLACEHOLDER_PATTERNS):
        return True
    return bool(_MEMBER_ACCESS_RE.match(stripped))


def scan_file_for_secrets(file_bytes: bytes) -> tuple[bool, str | None]:
    """Scan file content for credential patterns.

    Returns ``(rejected, reason)``. When rejected is True the upload must
    not be indexed; ``reason`` contains the detect-secrets detector label
    (e.g. ``"AWS Access Key"``, ``"Secret Keyword"``).

    Detector errors (malformed bytes, I/O failures) fail open — log and
    allow the upload rather than block it.
    """
    try:
        # detect-secrets scans a file path, so write to a temp file.
        # `/tmp` is tmpfs in the Docker image, so this is an in-memory
        # round-trip with no real disk I/O.
        with tempfile.NamedTemporaryFile(mode="wb", suffix=".txt", delete=False) as tf:
            tf.write(file_bytes)
            tmp_path = Path(tf.name)

        try:
            collection = SecretsCollection()
            with transient_settings({"plugins_used": _PLUGINS}):
                collection.scan_file(str(tmp_path))
        finally:
            tmp_path.unlink(missing_ok=True)
    except Exception:
        logger.exception("Secret scan failed; allowing file")
        return False, None

    for _filename, secret in collection:
        value = secret.secret_value or ""
        if _is_noise(value):
            logger.info(
                "Secret scanner: filtered placeholder match",
                extra={"type": secret.type, "value_len": len(value)},
            )
            continue
        logger.warning(
            "File rejected by secret scanner",
            extra={"type": secret.type, "value_len": len(value)},
        )
        return True, f"Potential secret detected: {secret.type}"

    return False, None

File: services/platform/convex/rag/lib/secret_scanner.ts
'use node';

/**
 * Pre-ingestion secret scanner.
 *
 * CONTRACT-FIDELITY NOTE: the Python implementation wrapped Yelp's
 * `detect-secrets` library (a large plugin pipeline + entropy detectors).
 * That library has no maintained JS equivalent, so this is a hand-rolled
 * detector that reproduces the behaviour the test suite pins:
 *
 * - specific-provider detectors (AWS access keys, PEM/OpenSSH private keys,
 *   JWTs);
 * - a keyword detector that flags a high-entropy / long literal assigned to a
 *   key/secret/password/token-shaped name;
 * - the same placeholder + member-access + templated-value noise filter the
 *   Python post-filter applied.
 *
 * It is deliberately narrower than `detect-secrets`; full parity will be
 * revisited against the live stack. Detector errors fail OPEN (log + allow).
 */

import { logger } from '../../lib/knowledge/logger';

export interface SecretScanResult {
  rejected: boolean;
  reason: string | null;
}

// Bareword placeholders the keyword detector would otherwise flag.
const PLACEHOLDER_VALUES: ReadonlySet<string> = new Set([
  '',
  'null',
  'none',
  'nil',
  'undefined',
  'true',
  'false',
  'redacted',
  'placeholder',
  'example',
  'changeme',
  'xxx',
  'xxxx',
  'xxxxxxxx',
  'tbd',
  'todo',
  'fixme',
  'n/a',
]);

const PLACEHOLDER_PATTERNS: RegExp[] = [
  // `your-api-key-here`, `my_secret`, `example-token`, etc.
  /^(?:your|my|sample|example|test|fake|dummy|demo)[-_].+/i,
  // `...-goes-here`, `...-placeholder`, `...-example`.
  /.+[-_](?:here|placeholder|example|sample|value|goes-here)$/i,
];

// Member-access chains (`config.apiKey`, `process.env.FOO`, `this.token`).
const MEMBER_ACCESS_RE = /^[A-Za-z_$][\w$]*(?:\.[A-Za-z_$][\w$]*)+$/;

// A masked value such as `********`.
const STAR_MASK_RE = /^\*+$/;

/** True when the detected value is a known placeholder, not a real secret. */
function isNoise(secretValue: string | null | undefined): boolean {
  if (!secretValue) {
    return true;
  }
  const stripped = secretValue.trim().replace(/^['"`]+|['"`]+$/g, '');
  if (!stripped) {
    return true;
  }
  const lowered = stripped.toLowerCase();
  if (PLACEHOLDER_VALUES.has(lowered)) {
    return true;
  }
  if (STAR_MASK_RE.test(stripped)) {
    return true;
  }
  if (PLACEHOLDER_PATTERNS.some((p) => p.test(lowered))) {
    return true;
  }
  // Templated values: `${env.TOKEN}`, `${process.env.KEY}`, `<foo>`, `{{x}}`.
  if (/\$\{.*\}|^<.+>$|\{\{.+\}\}/.test(stripped)) {
    return true;
  }
  return MEMBER_ACCESS_RE.test(stripped);
}

/** Shannon entropy of a string in bits per character. */
function shannonEntropy(value: string): number {
  if (value.length === 0) {
    return 0;
  }
  const counts = new Map<string, number>();
  for (const ch of value) {
    counts.set(ch, (counts.get(ch) ?? 0) + 1);
  }
  let entropy = 0;
  for (const count of counts.values()) {
    const p = count / value.length;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

interface Detector {
  type: string;
  detect: (text: string) => string | null;
  // Whether the post-detection noise filter applies. Specific-provider
  // detectors (AWS/PrivateKey/JWT) return a typed secret that must never be
  // noise-filtered — only the keyword detector's free-form value is.
  applyNoiseFilter: boolean;
}

const AWS_RE = /\b((?:AKIA|ASIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA)[A-Z0-9]{16})\b/;
const PEM_PRIVATE_KEY_RE =
  /-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----/;
const JWT_RE = /\beyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/;

// `name = "value"`, `name: value`, `name=value` for key-shaped names.
const KEYWORD_ASSIGN_RE =
  /(?:api[_-]?key|secret|password|passwd|pwd|token|auth)\s*[:=]\s*(.+)/gi;

const HEX_RE = /^[0-9a-fA-F]+$/;
const BASE64_RE = /^[A-Za-z0-9+/=_-]+$/;

const DETECTORS: Detector[] = [
  {
    type: 'AWS Access Key',
    applyNoiseFilter: false,
    detect: (text) => {
      const m = AWS_RE.exec(text);
      return m ? m[1] : null;
    },
  },
  {
    type: 'Private Key',
    applyNoiseFilter: false,
    detect: (text) => (PEM_PRIVATE_KEY_RE.test(text) ? 'PRIVATE KEY' : null),
  },
  {
    type: 'JSON Web Token',
    applyNoiseFilter: false,
    detect: (text) => {
      const m = JWT_RE.exec(text);
      return m ? m[0] : null;
    },
  },
  {
    type: 'Secret Keyword',
    applyNoiseFilter: true,
    detect: (text) => {
      KEYWORD_ASSIGN_RE.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = KEYWORD_ASSIGN_RE.exec(text)) !== null) {
        const rawValue = match[1].trim();
        const value = rawValue.replace(/^['"`]+|['"`]+$/g, '').trim();
        if (isNoise(value) || value.length < 12) {
          continue;
        }
        // High entropy OR long hex/base64 literal assigned to a key name.
        const isHex = HEX_RE.test(value) && value.length >= 16;
        const isBase64 =
          BASE64_RE.test(value) &&
          value.length >= 16 &&
          shannonEntropy(value) >= 3.5;
        if (isHex || isBase64) {
          return value;
        }
      }
      return null;
    },
  },
];

/**
 * Scan file content for credential patterns. Returns `{ rejected, reason }`.
 * When `rejected` is true the upload must not be indexed. Detector errors
 * fail open — log and allow.
 */
export function scanFileForSecrets(fileBytes: Uint8Array): SecretScanResult {
  let text: string;
  try {
    // Best-effort UTF-8 decode (non-fatal), mirroring detect-secrets' file read.
    text = new TextDecoder('utf-8', { fatal: false }).decode(fileBytes);
  } catch (err) {
    logger.error(
      `Secret scan failed; allowing file: ${err instanceof Error ? err.message : String(err)}`,
    );
    return { rejected: false, reason: null };
  }

  try {
    for (const detector of DETECTORS) {
      const value = detector.detect(text);
      if (value === null) {
        continue;
      }
      if (detector.applyNoiseFilter && isNoise(value)) {
        logger.info(
          `Secret scanner: filtered placeholder match (type=${detector.type})`,
        );
        continue;
      }
      logger.warn(
        `File rejected by secret scanner (type=${detector.type}, value_len=${value.length})`,
      );
      return {
        rejected: true,
        reason: `Potential secret detected: ${detector.type}`,
      };
    }
  } catch (err) {
    logger.error(
      `Secret scan failed; allowing file: ${err instanceof Error ? err.message : String(err)}`,
    );
    return { rejected: false, reason: null };
  }

  return { rejected: false, reason: null };
}
